File: presentation/src/main/java/com/github/chojmi/inspirations/presentation/model/gallery/GridAdapterUiModel.java
package com.github.chojmi.inspirations.presentation.model.gallery;

import android.support.annotation.Nullable;

import com.google.auto.value.AutoValue;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Observable;
import timber.log.Timber;

@AutoValue
public abstract class GridAdapterUiModel {
    public static GridAdapterUiModel create(Photo photo) {
        return new AutoValue_GridAdapterUiModel(photo, null, null);
    }

    public static List<GridAdapterUiModel> create(List<Photo> photos) {
        List<GridAdapterUiModel> uiModels = new ArrayList<>();
        Observable.fromIterable(photos)
                .map(GridAdapterUiModel::create)
                .toList()
                .subscribe(uiModels::addAll, Timber::e);
        return uiModels;
    }

    public static GridAdapterUiModel setFavs(GridAdapterUiModel model, PhotoFavs favs) {
        return new AutoValue_GridAdapterUiModel(model.getPhoto(), favs, model.getComments());
    }

    public static GridAdapterUiModel setComments(GridAdapterUiModel model, PhotoComments comments) {
        return new AutoValue_GridAdapterUiModel(model.getPhoto(), model.getFavs(), comments);
    }

    public abstract Photo getPhoto();

    @Nullable
    public abstract PhotoFavs getFavs();

    @Nullable
    public abstract PhotoComments getComments();
}

File: presentation/src/main/java/com/github/chojmi/inspirations/presentation/gallery/model/GridAdapterUiModel.kt
package com.github.chojmi.inspirations.presentation.gallery.model

import com.github.chojmi.inspirations.domain.entity.photos.PhotoInfoEntity
import com.github.chojmi.inspirations.domain.entity.photos.PhotoSizeEntity
import com.github.chojmi.inspirations.domain.entity.photos.PhotoSizeListEntity
import io.reactivex.Observable
import timber.log.Timber
import java.util.*

data class GridAdapterUiModel(val photo: PhotoWithAuthor, val favs: PhotoFavs?, val photoInfo: PhotoInfoEntity?,
                              val photoSizes: PhotoSizeListEntity<PhotoSizeEntity>?) {

    companion object {
        fun create(photo: PhotoWithAuthor): GridAdapterUiModel =
                GridAdapterUiModel(photo, null, null, null)

        fun create(photos: List<PhotoWithAuthor>): List<GridAdapterUiModel> {
            val uiModels = ArrayList<GridAdapterUiModel>()
            Observable.fromIterable(photos)
                    .map({ photo -> GridAdapterUiModel.Companion.create(photo) })
                    .toList()
                    .subscribe({ list -> uiModels.addAll(list) }, { t -> Timber.d(t) })
            return uiModels
        }

        fun setFavs(model: GridAdapterUiModel, favs: PhotoFavs): GridAdapterUiModel =
                GridAdapterUiModel(model.photo, favs, model.photoInfo, model.photoSizes)

        fun setPhotoInfo(model: GridAdapterUiModel, photoInfo: PhotoInfoEntity): GridAdapterUiModel =
                GridAdapterUiModel(model.photo, model.favs, photoInfo, model.photoSizes)

        fun setPhotoSizes(model: GridAdapterUiModel, sizes: PhotoSizeListEntity<PhotoSizeEntity>) =
                GridAdapterUiModel(model.photo, model.favs, model.photoInfo, sizes)
    }
}
